src/utils/MasterFisher.ts

```diff
@@ -0,0 +1,104 @@
+ * being exported. When the MasterFisher fishes, it fishes in the package first, then the tank, and
+ * then the FHIRDefinitions.  This essentially prefers local definitions first (when there are naming
+ * clashes).
+ *
+ * The MasterFisher also uses its knowledge of these Fishable locations to do some necessary magic.
+ * For instance, the FSHTank has no knowledge of FHIRDefinitions, so it cannot provide the correct
+ * sdType for its metadata results.  When the MasterFisher detects this has happened, it uses the
+ * other Fishable locations to determine the proper sdType (even for an item that currently exists
+ * only in the tank).
+ */
+export class MasterFisher implements Fishable {
+  constructor(public tank?: FSHTank, public fhir?: FHIRDefinitions, public pkg?: Package) {}
+
+  /**
+   * Searches for the FHIR JSON by name/id/url.  It will first search through the local package
+   * (which contains FHIR artifacts exported so far), then through the external FHIR definitions.
+   * @param {string} item - the item name/id/url to fish for
+   * @param types - the allowable types to fish for
+   */
+  fishForFHIR(item: string, ...types: Type[]): any | undefined {
+    // Resolve the alias if necessary
+    item = this.tank?.resolveAlias(item) ?? item;
+
+    // First check for it in the package
+    let result = this.pkg?.fishForFHIR(item, ...types);
+    if (result == null) {
+      // If it is in the tank, return undefined. We don't want to return the external FHIR
+      // definition, even if it exists -- because it won't match what is in the tank.  This
+      // ensures consistency between the outputs of fishForFHIR and fishForMetadata.
+      if (this.tank?.fish(item, ...types)) {
+        return;
+      }
+      result = this.fhir?.fishForFHIR(item, ...types);
+    }
+    return result;
+  }
+
+  /**
+   * Searches for the Metadata associated with the passed in name/id/url.  It will first search
+   * through the local package (which contains FHIR artifacts exported so far), then through the
+   * tank, then through the external FHIR definitions. This function is useful because it gets
+   * commonly used information without having to force an export. This helps to reduce the risk
+   * of circular dependencies causing problems.
+   * @param item - the item/name/id url to fish for
+   * @param types - the allowable types to fish for
+   */
+  fishForMetadata(item: string, ...types: Type[]): Metadata {
+    // Resolve the alias if necessary
+    item = this.tank?.resolveAlias(item) ?? item;
+
+    const fishables = [this.pkg, this.tank, this.fhir].filter(f => f != null);
+    for (const fishable of fishables) {
+      const result = fishable.fishForMetadata(item, ...types);
+      if (result) {
+        // If it came from the tank, we need to get the sdType because the tank doesn't know.
+        if (fishable instanceof FSHTank) {
+          result.sdType = this.findSdType(result, types, fishables);
```

Comment:
I tried deleting all of the cached definitions in `testdefs/package`, and there is still one test that fails. The test is 'should apply a correct OnlyRule on a non-reference FSHy choice' in StructureDefinitionExporter.test.ts. If you want to fix the test itself, you can add a `fisher.fishForStructureDefinition('Quantity');` line to it before exporting in that test.

But the test led me to here. In the case of the test, `MySpecialQuantity` is a child of `Quantity`, but `Quantity` is not present in the definitions, and we are constraining a `value[x]` element to only be `MySpecialQuantity`. Since `Quantity` is not defined, `findSdType` returns `undefined`, and `result` is returned with `result.sdType = undefined`. So I wonder if we should still return `result` if `fishable` is a `FSHTank` but we don't actually find an `sdType`.

Reply:
Thanks for the thorough review, @ngfreiter!  In the case exercised by the test, if this were a real run of sushi, there would be much more troublesome errors than a null `sdType` -- as the export of `MySpecialQuantity` would fail completely (since its parent is missing).  In general, I think the only way this will happen is when there are FSHy profiles/extensions that have missing parents -- which will always present itself as a bigger problem.

That being the case, it's not possible that a null `sdType` would somehow go unnoticed -- there will always be bigger/badder errors alongside it; and once the author fixes those the `sdType` issue is also fixed.  So... I think I'm ok with it as-is, which is basically a "best effort" / "return as much as you can" approach.  It's actually consistent with our overall approach which is that even if we run into an error in processing a rule on a profile, we don't toss the whole profile -- we do our best to export as much of it as we can.

Don't let me fool you though.  It's not like I thought of this specific case and intentionally designed it that way -- so it's good that you brought it up and forced the conversation.  Let me know if you disagree with my thoughts on it.

Reply:
Yup that's totally fine by me! I can't imagine how it would be an issue without other errors finding it, so I'm good.

Reply:
I just pushed up a commit to add the fishing statement.